src/cluster_resources.rs

```diff
@@ -0,0 +1,393 @@
+        if let Some(actual_content) = labels.get(label) {
+            if expected_content == actual_content {
+                Ok(())
+            } else {
+                Err(Error::UnexpectedLabelContent {
+                    label,
+                    expected_content: expected_content.into(),
+                    actual_content: actual_content.into(),
+                })
+            }
+        } else {
+            Err(Error::MissingLabel { label })
+        }
+    }
+
+    /// Finalizes the cluster creation.
+    ///
+    /// All orphaned resources, i.e. resources which are labelled as if they belong to this cluster
+    /// instance but were not added to the cluster resources, are deleted.
+    ///
+    /// The following resource labels are compared:
+    /// * `app.kubernetes.io/instance`
+    /// * `app.kubernetes.io/managed-by`
+    /// * `app.kubernetes.io/name`
+    ///
+    /// # Arguments
+    ///
+    /// * `client` - The client which is used to access Kubernetes
+    pub async fn finalize(self, client: &Client) -> OperatorResult<()> {
+        self.delete_orphaned_resources::<Service>(client).await?;
+        self.delete_orphaned_resources::<StatefulSet>(client)
+            .await?;
+        self.delete_orphaned_resources::<DaemonSet>(client).await?;
+        self.delete_orphaned_resources::<ConfigMap>(client).await?;
+
+        Ok(())
+    }
+
+    /// Deletes all deployed resources which are labelled as if they belong to this cluster
+    /// instance but are not contained in the given list.
+    ///
+    /// If it is forbidden to list the resources of the given kind then it is assumed that the
+    /// caller is not in charge of these resources, the deletion is skipped, and no error is
+    /// returned.
+    ///
+    /// # Arguments
+    ///
+    /// * `client` - The client which is used to access Kubernetes
+    async fn delete_orphaned_resources<T: ClusterResource>(
+        &self,
+        client: &Client,
+    ) -> OperatorResult<()> {
+        match self.list_deployed_cluster_resources::<T>(client).await {
+            Ok(deployed_cluster_resources) => {
+                let orphaned_resources = deployed_cluster_resources
+                    .into_iter()
+                    .filter(|r| !self.resources.contains(&r.into()))
+                    .collect::<Vec<_>>();
+
+                if !orphaned_resources.is_empty() {
+                    info!(
+                        "Deleting orphaned {}: {}",
+                        T::plural(&()),
+                        ClusterResources::print_resources(&orphaned_resources),
+                    );
+                    for resource in orphaned_resources.iter() {
+                        client.delete(resource).await?;
+                    }
+                }
+
+                Ok(())
+            }
+            Err(Error::KubeError {
+                source: kube::Error::Api(ErrorResponse { code: 403, .. }),
+            }) => {
+                debug!(
```

Comment:
Should this maybe be at the warn level? Usually this will mean something is messed up, no?

Reply:
This case occurs for all resources which are not used by the operator. Most operators do not use DaemonSets and then this "error" occurs which can be ignored. The log message is shown on every `reconcile` call, so it should have a log level below the default one which is probably "info".

The error could be avoided if the user would have an option to enter the used resource kinds in `ClusterResources::new` but this is cumbersome and error-prone if the operator is extended.

Reply:
ok